app/assets/javascripts/dashboard/dataset/team_selection_component.js

```diff
@@ -0,0 +1,74 @@
+// @flow
+import _ from "lodash";
+import * as React from "react";
+import { Select } from "antd";
+import { getEditableTeams } from "admin/admin_rest_api";
+import type { APITeamType } from "admin/api_flow_types";
+
+const { Option } = Select;
+
+type Props = {
+  value: Array<APITeamType>,
+  onChange: (value: Array<APITeamType>) => void,
+  mode?: ?string,
+};
+
+type State = {
+  editableTeams: Array<APITeamType>,
+  allowedTeams: Array<APITeamType>,
+};
+
+class TeamSelectionComponent extends React.PureComponent<Props, State> {
+  state = {
+    editableTeams: [],
+    allowedTeams: this.props.value ? this.props.value : [],
+  };
+
+  componentDidMount() {
+    this.fetchData();
+  }
+
+  componentWillReceiveProps(newProps: Props) {
+    this.setState({
+      allowedTeams: newProps.value,
+    });
+  }
+
+  async fetchData() {
+    const editableTeams = await getEditableTeams();
+    this.setState({
+      editableTeams,
+    });
+  }
+
+
+  onSelectTeams = (selectedTeamIds: Array<string>) => {
+    const allTeams  = this.getAllTeams();
+    const allowedTeams = _.compact(selectedTeamIds.map(id => allTeams.find(t => t.id === id)))
+    this.props.onChange(allowedTeams);
+    this.setState({
+      allowedTeams,
+    });
+  };
+
+  getAllTeams = () => _.uniqBy(this.state.editableTeams.concat(this.state.allowedTeams), t => t.id)
```

Comment:
Let's be extra fancy and use `_.unionBy` like this: `getAllTeams = () => _.unionBy(this.state.editableTeams, this.state.allowedTeams, t => t.id)`